### PR TITLE
RouterOS QoS Support

### PR DESCRIPTION
**Pull request by eskyuu**
Adds support for both Simple Queues (device level) and Queue Trees (port level) QoS Queues for Mikrotik devices.

Device level graphs (not linked to a port.  It would show a tree if Mikrotik provided parent information in SNMP) :
image

Port level graphs:
image

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [ ] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [ ] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by eskyuu** [comment]
Feature was requested here:
https://community.librenms.org/t/feature-request-support-routeros-queues/986

**Comment by eskyuu** [comment]
The test data is blank.  Not sure if this is because I've done all the discovery and polling work inside a function (to avoid making any global variables), or if it's because it creates components.

**Comment by laf** [comment]
I'm not a fan of the generic component table, it makes things like alert rule building harder. Does this need it's own table really?

**Comment by eskyuu** [comment]
Fair comment - I was thinking about alerts and whether I could add some more prefs to save things like rate in/out and drop in/out.  I've used the Cisco CBQoS as the template for how to store data, assuming this was a preferred way to do things.

Do you think a generic "QoS" model works?  I was thinking that this generic model could store delta in/out for both transmitted and dropped bytes, as well as a calculated column for dropped percent.  These would be the things that can be alerted on.

There would still be the ability to poll additional fields in the poller, but they would just not be reportable without a separate model (or maybe extending the QoS model?).

I should be able to get the conversion done relatively quickly, but don't want to waste my time if there is either an existing model that already exists that I should use, or if a generic QoS model is a bad idea and won't get merged.  I'm assuming that as a general rule we don't want to end up with a table for each specific graph type (os and model specific).

**Comment by eskyuu** [comment]
I was also originally thinking of having a way to migrate the Cisco CBQoS stuff to be a second example in the new schema.

**Comment by eskyuu** [comment]
How does this version look?  I need to add a space between the policies and the graphs, but other than that I think the functionality is correct.

**Comment by eskyuu** [comment]
And is it worth adding any documentation around how to add a new QoS type?

**Comment by eskyuu** [comment]
and I think I need a settings page to allow individual QoS queues to be disabled

**Comment by murrant** [comment]
over time, we have been combining vendor specific tables into more generic ones with fields that work for both.

Closest I can get to docs is to look at other pull requests:

**Comment by murrant** [comment]
Generally try to move from vendor specific modules to generic modules with a table structure that can support both.

The Slas module is a recent example. There are probably others.
https://github.com/librenms/librenms/pull/12799/files
https://github.com/librenms/librenms/pull/14288/files

**Comment by eskyuu** [comment]
Thanks - that is what I was looking for to work out the correct current way to do polling and discovery.

How does it look now?

**Comment by eskyuu** [comment]
The tests are failing because there is a "last_polled" field which always shows the current timestamp.  What is the workaround for this?

File: database/migrations/2024_10_20_154356_create_qos_table.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            $table->bigInteger('last_packets_in')->nullable()->comment('Last polled counter for input packets');
+            $table->bigInteger('last_packets_out')->nullable()->comment('Last polled counter for output packets');
+            $table->bigInteger('packets_in_rate')->nullable()->comment('Output rate for packets');
+            $table->bigInteger('packets_out_rate')->nullable()->comment('Input rate for packets');
+            $table->bigInteger('last_packets_drop_in')->nullable()->comment('Last polled counter for input dropped packets');
+            $table->bigInteger('last_packets_drop_out')->nullable()->comment('Last polled counter for output dropped packets');
+            $table->bigInteger('packets_drop_in_rate')->nullable()->comment('Output rate for dropped packets');
+            $table->bigInteger('packets_drop_out_rate')->nullable()->comment('Input rate for dropped packets');
+            $table->bigInteger('bytes_drop_in_pct')->storedAs('CASE WHEN bytes_in_rate = 0 THEN 0 ELSE 100 * bytes_drop_in_rate / bytes_in_rate END')->comment('Percentage of input bytes dropped');
```
Comment (murrant) [inline database/migrations/2024_10_20_154356_create_qos_table.php:47]: I'm not sure about this.  Adding logic at the database layer has bitten us in the ass a lot.

**Comment by murrant** [comment]
Looking pretty good over all.  I would like to see another OS pulled in to make sure the database structure makes sense.

File: database/migrations/2024_10_20_154356_create_qos_table.php (hunk 1 above)
Comment (eskyuu) [reply database/migrations/2024_10_20_154356_create_qos_table.php]: Logic has been moved to the observer code

**Comment by eskyuu** [comment]
I was going to have a shot at moving Cisco cbqos to this new system in the next release cycle.  I just don't have easy access to a dev system that can see a Cisco router, and I'm already juggling a lot of patches on our production and staging systems.  I'm hoping to simplify that environment a lot more with the next release (assuming other PRs are merged), which will allow me to play around with this QoS stuff a bit easier.

**Comment by eskyuu** [comment]
I've implemented the Cisco CBQoS, which did prompt adding a non-DB field to the schema to pass values from the OS poller function to the QoS Module that we are keeping for backwards compatibility with the existing functionality.  It also prompted adding a tooltip field to store additional data about the queue that we don't want to display on the web page.

My concern (and the reason I didn't include the Cisco stuff to begin with) is that this change now has the very real potential to change existing deployments.

**Comment by laf** [comment]
Great work as always @eskyuu.

I'm personally happy to merge this in after the next release.

**Comment by eskyuu** [comment]
@laf Are you still happy to approve this?

**Comment by murrant** [comment]
This looks good to me @eskyuu 

Sorry, my PR to remove cisco-voice means you need to rebase this.

**Review by laf — APPROVED** [review]
LGTM

**Comment by M4rlus** [comment]
Hey there @eskyuu @murrant ,

I just noticed an issue with the traffic calculation for RouterOS/MikroTik devices that seems related to this code.

On my MikroTik devices, the traffic values reported by LibreNMS are always off by a wierd factor. I’ve attached two screenshots: one showing the actual traffic on the uplink interface, and one showing the corresponding parent queue. When comparing both, LibreNMS qos graph consistently displays traffic that is higher than what RouterOS reports.

This issue occurs on both RouterOS 6.x and 7.x. I have access to multiple CCR devices with queues, and the problem is consistent across all of them.

I’m a bit unsure whether it’s worth opening a separate issue for this, but I wanted to mention it here in case it helps.

Let me know if more data or testing would help.

Queue simple traffic:
image


Actual port speed:
image

**Comment by eskyuu** [comment]
This would be best raised as a new issue ticket.  Please make sure you specify the type of queue.  I just checked and queue trees look good, but simple queues didn't.  It's possible that Mikrotik are not returning bytes in their SNMP oids (mtxrQueueSimpleBytesOut and mtxrQueueSimpleBytesIn), which would make the LibreNMS calculation wrong.  This would require contacting Mikrorik to get confirmation of what their number means.

I changed the multiplier in includes/html/graphs/device/qos_routeros_simple_traffic.inc.php to 1, and my graphs looked better, but I'm not sure that would be the case for the numbers you're seeing.

**Comment by andhangkurniawan** [comment]
RouterOS QoS not working 
Screenshot 2025-12-07 at 23-22-48 core murdocktechnology net MRDK NMS
./validate.php
===========================================
Component | Version
--------- | -------
LibreNMS  | 25.12.0-dev.48+5e3e87602 (2025-12-07T03:29:55+07:00)
DB Schema | 2025_11_25_102228_increase_increase_port_descr_speed_length (361)
PHP       | 8.3.6
Python    | 3.12.3
Database  | MariaDB 10.11.13-MariaDB-0ubuntu0.24.04.1
RRDTool   | 1.7.2
SNMP      | 5.9.4.pre2
===========================================

[OK]    Composer Version: 2.9.2
[OK]    Dependencies up-to-date.
[OK]    Database Connected
[OK]    Database Schema is current
[OK]    SQL Server meets minimum requirements
[OK]    lower_case_table_names is enabled
[OK]    MySQL engine is optimal
[OK]    Database and column collations are correct
[OK]    Database schema correct
[OK]    MySQL and PHP time match
[OK]    Active pollers found
[OK]    Dispatcher Service not detected
[OK]    Locks are functional
[OK]    Python poller wrapper is polling
[OK]    Redis is unavailable
[OK]    rrd_dir is writable
[OK]    All 1849 RRD files have the correct step.
[OK]    rrdtool version ok

librenms.log
[2025-12-07T23:29:19][CRITICAL] Exception: Error Call to a member function getSpeeds() on null @ /opt/librenms/includes/html/graphs/generic_data.inc.php:37
#0 /opt/librenms/includes/html/graphs/device/qos_routeros_simple_traffic.inc.php(11): require()
#1 /opt/librenms/includes/html/graphs/graph.inc.php(50): include('...')
#2 /opt/librenms/html/graph.php(29): require('...')
#3 {main}

**Comment by murrant** [comment]
If you have issues, create a new issue and reference this PR....

**Pull request merged**